### ISSUE TITLE
Arithmetic assignment operators gives wrong answers

### ISSUE THREAD

**Issue by calebh** [pytorch/pytorch#7731]
## Issue description

Using the +=, -=, *=, /= operators gives the wrong answer in some cases without any warning. The matrix multiply assignment operator @= seems to work fine. I have tested in both PyTorch 0.3.0.post4 and PyTorch 0.4.0 (the latest version at the time of this issue creation) on the CPU. I have not tested whether this bug happens with CUDA.

## Code example

I was able to produce a minimal case of this bug:

```
import torch

## Addition
x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

t1 *= t2
print("Incorrect answer")
print(t1)

x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

print("Correct answer")
print(t1 * t2)

## Subtraction
x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

t1 -= t2
print("Incorrect answer")
print(t1)

x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

print("Correct answer")
print(t1 - t2)

## Multiplication
x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

t1 *= t2
print("Incorrect answer")
print(t1)

x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

print("Correct answer")
print(t1 * t2)

## Division
x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

t1 /= t2
print("Incorrect answer")
print(t1)

x = torch.Tensor([[1],[2],[3]])
t1 = x.expand(3, 3)
t2 = x.t().expand(3, 3)

print("Correct answer")
print(t1 / t2)

```

## Output
```
Incorrect answer
tensor([[ 6.0000e+00,  6.0000e+00,  6.0000e+00],
        [ 4.3200e+02,  4.3200e+02,  4.3200e+02],
        [ 6.0466e+07,  6.0466e+07,  6.0466e+07]])
Correct answer
tensor([[ 1.,  2.,  3.],
        [ 2.,  4.,  6.],
        [ 3.,  6.,  9.]])
Incorrect answer
tensor([[-5., -5., -5.],
        [-3., -3., -3.],
        [ 0.,  0.,  0.]])
Correct answer
tensor([[ 0., -1., -2.],
        [ 1.,  0., -1.],
        [ 2.,  1.,  0.]])
Incorrect answer
tensor([[ 6.0000e+00,  6.0000e+00,  6.0000e+00],
        [ 4.3200e+02,  4.3200e+02,  4.3200e+02],
        [ 6.0466e+07,  6.0466e+07,  6.0466e+07]])
Correct answer
tensor([[ 1.,  2.,  3.],
        [ 2.,  4.,  6.],
        [ 3.,  6.,  9.]])
Incorrect answer
tensor([[ 0.1667,  0.1667,  0.1667],
        [ 0.3333,  0.3333,  0.3333],
        [ 1.0000,  1.0000,  1.0000]])
Correct answer
tensor([[ 1.0000,  0.5000,  0.3333],
        [ 2.0000,  1.0000,  0.6667],
        [ 3.0000,  1.5000,  1.0000]])

```

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 17.10
GCC version: (Ubuntu 7.2.0-8ubuntu3.2) 7.2.0
CMake version: version 3.9.1

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.14.3)
[pip3] torch (0.4.0)
[conda] Could not collect

**Comment by zou3519** [comment]
Thanks for the report, @calebh. When tensors share the same storage, in-place operations are incorrect. You should generally avoid in-place operations because they can have other bad side effects, like making backward computation impossible (autograd will throw an error if that is the case though).

Ideally we'd throw an error or warn if the tensors involved in an in-place operation have overlapping storage, but determining that fully is equivalent to finding a solution to a linear diophantine equation in `n` variables, where `n` is proportional to the number of dimensions in the input tensors.

**Comment by SsnL** [comment]
Yes this is a difficult problem. The issue was previously raised in https://github.com/pytorch/pytorch/issues/5272

**Comment by calebh** [comment]
If fixing this issue is impossible, then there should at least be a big warning in the documentation. I've spent a lot of time in the documentation and I haven't seen any warning about these operators. Another approach would be to disable these operators completely unless a global flag has been set which tells PyTorch "it's okay, I know what I'm doing." This would allow power users to still get the performance they want without sacrificing safety for new people.

This issue calls into question a lot of the work that I've done in the past couple months. I occasionally use the arithmetic assignment operators, and I was assuming that += was just syntax sugar for the + operation. I wouldn't want someone else to go through this same trouble.

**Comment by zou3519** [comment]
It's not impossible to fix, but it is difficult. I agree that having a big warning in the docs is good, though.

**Comment by ezyang** [comment]
> but determining that fully is equivalent to finding a solution to a linear diophantine equation in n variables, where n is proportional to the number of dimensions in the input tensors.

Even if this is true, the examples calebh posted are very simple and would be caught by the current test we have in PyTorch. We should apply this test.

**Comment by fmassa** [comment]
@calebh just to reiterate, `+=` is NOT syntactic sugar for `x = x + ...`. The same holds true for python lists for example, while one mutates elements in place, while the other creates a new list.
